### ISSUE TITLE
Combine network object registration and function registration on those objects

### ISSUE THREAD

**Issue by lyonsil** [paranext/paranext-core#151]
Currently you have to send a message to register a network object, then another message for each function to make available on that object.  This could all just be a single request where the list of functions is provided along with the name of the object to register.

**Comment by tjcouch-sil** [comment]
Actually, the network object `.function` request is the actual function call on the network object. There is no discovering what functions are on a network object right now (ingestors just request `.function` and put a function name as the first element in the contents array in hopes that the network object has that function), but the contents of the `.get` function are an array so that, in the future, network objects can announce what functions they have.

**Comment by lyonsil** [comment]
Right, but you still have to send a separate network message for every function being registered according to https://docs.google.com/document/d/1uPkrxoa4WjUwdBZLetjHf3F49LHqbni8STn-Dm8zu5k:

```
A client registers two request handlers:
{"type":"request","requestType":"server:registerRequest","requestId":17,"senderId":6,"contents":["object:hello-someone.greetings-data.get",6]}
{"type":"request","requestType":"server:registerRequest","requestId":18,
"senderId":6,"contents":["object:hello-someone.greetings-data.function",6]}
```

This ticket is suggesting there should be a single registration message for a network object.  One possibility of how this could look is
```
{"type":"request","requestType":"registerNetworkObject","requestTarget":"server","requestId":5,"contents":["calculator","add","subtract","multiply","divide"],"senderId":2}
```

Discovery is something to consider but not intended to be part of this particular issue.

**Comment by lyonsil** [comment]
#148, #149, and #150 address discovery in some fashion.

**Comment by tjcouch-sil** [comment]
> Right, but you still have to send a separate network message for every function being registered according to https://docs.google.com/document/d/1uPkrxoa4WjUwdBZLetjHf3F49LHqbni8STn-Dm8zu5k:
> 
> ```
> A client registers two request handlers:
> {"type":"request","requestType":"server:registerRequest","requestId":17,"senderId":6,"contents":["object:hello-someone.greetings-data.get",6]}
> {"type":"request","requestType":"server:registerRequest","requestId":18,
> "senderId":6,"contents":["object:hello-someone.greetings-data.function",6]}
> ```
> 
> This ticket is suggesting there should be a single registration message for a network object. One possibility of how this could look is
> 
> ```
> {"type":"request","requestType":"registerNetworkObject","requestTarget":"server","requestId":5,"contents":["calculator","add","subtract","multiply","divide"],"senderId":2}
> ```
> 
> Discovery is something to consider but not intended to be part of this particular issue.

I think I need to clarify a bit: every network object only registers two request handlers. One request handler is `.get`, which others can use to know that the network object exists. The other request handler is `.function`, a single request handler that routes any function another process asks to run to that object to run its function. The one single `.function` request handler can handle `networkObject.doStuff('things', 'apples')`, `networkObject.getMyData(5)`, `networkObject.destroyTheWorld()`, etc. Each of these function calls from a remote process turns into a request with type `.function`. So let's say our example network object has the id `network-object`. The calls to these three functions would look like this:

```json
{"type":"request","requestType":"object:network-object.function","requestId":26,"senderId":4,"contents":["doStuff","things","apples"]}
```
```json
{"type":"request","requestType":"object:network-object.function","requestId":27,"senderId":4,"contents":["getMyData",5]}
```
```json
{"type":"request","requestType":"object:network-object.function","requestId":28,"senderId":4,"contents":["destroyTheWorld"]}
```

Note that each function runs through the same request handler, `object:network-object.function`. We are not changing the `.function` part based on which function we are running; the first element in the contents array is the name of the function to run. The subsequent elements in the array are the arguments passed into that function.

Yes, it is strange for one network object to register two request handlers, `.get` and `.function`, that do different things. However, this was the path of least resistance to get to having network objects when layering over the underlying api features I already had (network objects are simply a nice abstraction over requests). Maybe we could rework all of the messages such that network objects are the primary feature instead of requests, and we wouldn't have to register two request handlers for one network object. However, for now, we have only two request handlers per network object, no more.

**Comment by lyonsil** [comment]
#1200 would change how this works. Presumably this is still possible if we add a new network endpoint that takes the name of a network object and all the functions included with it. However that would still mean all of the individual functions need individual method names to call via JSON-RPC.
